src/charm.py

```diff
@@ -0,0 +1,208 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+#
+# Learn more at: https://juju.is/docs/sdk
+
+"""Charm definition and helpers."""
+
+import logging
+import os
+
+from charms.nginx_ingress_integrator.v0.ingress import IngressRequires
+from jinja2 import Environment, FileSystemLoader
+from ops import framework, main
+from ops.charm import CharmBase
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
+
+from log import log_event_handler
+
+# Log messages can be retrieved using juju debug-log
+logger = logging.getLogger(__name__)
+
+
+def render(template_name, context):
+    """Render the template with the given name using the given context dict.
+
+    Args:
+        template_name: File name to read the template from.
+        context: Dict used for rendering.
+
+    Returns:
+        A dict containing the rendered template.
+    """
+    charm_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir))
+    loader = FileSystemLoader(os.path.join(charm_dir, "templates"))
+    return Environment(loader=loader).get_template(template_name).render(**context)
+
+
+class TemporalUiK8SOperatorCharm(CharmBase):
+    """Temporal UI charm.
+
+    Attrs:
+        _state: used to store data that is persisted across invocations.
+        external_hostname: DNS listing used for external connections.
+    """
+
+    _state = framework.StoredState()
+
+    @property
+    def external_hostname(self):
+        """Return the DNS listing used for external connections."""
+        return self.config["external-hostname"] or self.app.name
+
+    def __init__(self, *args):
+        """Construct.
+
+        Args:
+            args: Ignore.
+        """
+        super().__init__(*args)
+        self.name = "temporal-ui"
+
+        # Handle basic charm lifecycle.
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.temporal_ui_pebble_ready, self._on_temporal_ui_pebble_ready)
+
+        # Handle ui:temporal relation.
+        self._state.set_default(server_status="blocked")
+        self.framework.observe(self.on.ui_relation_joined, self._on_ui_relation_joined)
+        self.framework.observe(self.on.ui_relation_changed, self._on_ui_relation_changed)
+        self.framework.observe(self.on.ui_relation_broken, self._on_ui_relation_broken)
+
+        # Handle Ingress.
+        self.ingress = IngressRequires(
+            self,
+            {
+                "service-hostname": self.external_hostname,
+                "service-name": self.app.name,
+                "service-port": 8080,
+            },
+        )
+
+    @log_event_handler(logger)
+    def _on_install(self, event):
+        """Install temporal UI tools.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self.unit.status = MaintenanceStatus("installing temporal ui tools")
```

Comment:
According to the [charm's lifecycle](https://discourse.charmhub.io/t/a-charms-life/5938), the charm goes through the install state when juju is done provisioning the unit. This is shown in the charm's status momentarily until it is done deploying. Once done, it will move start the server and show the next status (e.g. "waiting for relation")